File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/filter/service/ServiceBySecretForeignKeys.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.rest.v1.dto.filter.service;


import com.google.common.base.Objects;
import org.taktik.icure.entities.embed.Service;
import org.taktik.icure.services.external.rest.handlers.JsonPolymorphismRoot;
import org.taktik.icure.services.external.rest.v1.dto.filter.Filter;

import java.util.Set;

@JsonPolymorphismRoot(Filter.class)
public class ServiceBySecretForeignKeys extends Filter<Service> implements org.taktik.icure.dto.filter.service.ServiceBySecretForeignKeys {
	String healthcarePartyId;
    Set<String> patientSecretForeignKeys;

	public ServiceBySecretForeignKeys() {
	}

	public ServiceBySecretForeignKeys(String healthcarePartyId, Set<String> patientSecretForeignKeys) {
		this.healthcarePartyId = healthcarePartyId;
		this.patientSecretForeignKeys = patientSecretForeignKeys;
	}

	@Override
	public String getHealthcarePartyId() {
		return healthcarePartyId;
	}

	public void setHealthcarePartyId(String healthcarePartyId) {
		this.healthcarePartyId = healthcarePartyId;
	}

	@Override
	public Set<String> getPatientSecretForeignKeys() {
		return patientSecretForeignKeys;
	}

	public void setPatientSecretForeignKey(Set<String> patientSecretForeignKeys) {
		this.patientSecretForeignKeys = patientSecretForeignKeys;
	}

    public void setPatientSecretForeignKeys(Set<String> patientSecretForeignKeys) {
        this.patientSecretForeignKeys = patientSecretForeignKeys;
    }

    @Override
	public boolean matches(Service item) {
		return patientSecretForeignKeys != null && !patientSecretForeignKeys.isEmpty() && (healthcarePartyId == null || item.getDelegations().keySet().contains(healthcarePartyId))
				&& (item.getSecretForeignKeys() != null && item.getSecretForeignKeys().stream().anyMatch(sfk -> patientSecretForeignKeys.contains(sfk)));
	}
}

File: src/main/kotlin/org/taktik/icure/domain/filter/impl/service/ServiceBySecretForeignKeys.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.domain.filter.impl.service

import com.github.pozo.KotlinBuilder
import org.taktik.icure.domain.filter.AbstractFilter
import org.taktik.icure.entities.embed.Service

@KotlinBuilder
data class ServiceBySecretForeignKeys(
        override val desc: String? = null,
        override val healthcarePartyId: String? = null,
        override val patientSecretForeignKeys: Set<String>
) : AbstractFilter<Service>, org.taktik.icure.domain.filter.service.ServiceBySecretForeignKeys {
    override fun matches(item: Service): Boolean {
        return (!patientSecretForeignKeys.isEmpty() && (healthcarePartyId == null || item.delegations.keys.contains(healthcarePartyId))
                && (item.secretForeignKeys?.any { sfk: String? -> patientSecretForeignKeys.contains(sfk) }) == true)
    }
}
